src/main/java/com/powsybl/openloadflow/AbstractAcOuterLoopConfig.java

```diff
@@ -0,0 +1,107 @@
+                                                                    parametersExt.getMaxPlausibleTargetVoltage()));
+        }
+        return Optional.empty();
+    }
+
+    protected static Optional<AcOuterLoop> createMonitoringVoltageOuterLoop(OpenLoadFlowParameters parametersExt) {
+        if (parametersExt.isSvcVoltageMonitoring()) {
+            return Optional.of(new MonitoringVoltageOuterLoop());
+        }
+        return Optional.empty();
+    }
+
+    protected static Optional<AcOuterLoop> createTransformerVoltageControlOuterLoop(LoadFlowParameters parameters, OpenLoadFlowParameters.TransformerVoltageControlMode controlMode,
+                                                                                    int incrementalTransformerVoltageControlOuterLoopMaxTapShift) {
+        if (parameters.isTransformerVoltageControlOn()) {
+            AcOuterLoop outerLoop = switch (controlMode) {
+                case WITH_GENERATOR_VOLTAGE_CONTROL -> new SimpleTransformerVoltageControlOuterLoop();
+                case AFTER_GENERATOR_VOLTAGE_CONTROL -> new TransformerVoltageControlOuterLoop();
+                case INCREMENTAL_VOLTAGE_CONTROL -> new IncrementalTransformerVoltageControlOuterLoop(incrementalTransformerVoltageControlOuterLoopMaxTapShift);
+            };
+            return Optional.of(outerLoop);
+        }
+        return Optional.empty();
+    }
+
+    protected static Optional<AcOuterLoop> createTransformerVoltageControlOuterLoop(LoadFlowParameters parameters, OpenLoadFlowParameters parametersExt) {
+        return createTransformerVoltageControlOuterLoop(parameters,
+                                                        parametersExt.getTransformerVoltageControlMode(),
+                                                        parametersExt.getIncrementalTransformerVoltageControlOuterLoopMaxTapShift());
+    }
+
+    protected static Optional<AcOuterLoop> createShuntVoltageControlOuterLoop(LoadFlowParameters parameters, OpenLoadFlowParameters.ShuntVoltageControlMode controlMode) {
+        if (parameters.isShuntCompensatorVoltageControlOn()) {
+            AcOuterLoop outerLoop = switch (controlMode) {
+                case WITH_GENERATOR_VOLTAGE_CONTROL -> new ShuntVoltageControlOuterLoop();
+                case INCREMENTAL_VOLTAGE_CONTROL -> new IncrementalShuntVoltageControlOuterLoop();
+            };
+            return Optional.of(outerLoop);
+        }
+        return Optional.empty();
+    }
+
+    protected static Optional<AcOuterLoop> createShuntVoltageControlOuterLoop(LoadFlowParameters parameters, OpenLoadFlowParameters parametersExt) {
+        return createShuntVoltageControlOuterLoop(parameters, parametersExt.getShuntVoltageControlMode());
+    }
+
+    protected static Optional<AcOuterLoop> createPhaseShifterControlOuterLoop(LoadFlowParameters parameters, OpenLoadFlowParameters.PhaseShifterControlMode controlMode) {
```

Comment:
In general, I don't like the use of phase shifter inside open loadflow as we have chosen phase control.

Reply:
Changed